File: shared/security/path-validator.ts
// shared/security/path-validator.ts
import * as path from "path";
import * as fs from "fs";

/**
 * Options for path validation
 */
export interface PathValidationOptions {
  /** Base directory - paths must resolve within this directory */
  baseDir: string;
  /** Allow symlinks (default: false for security) */
  allowSymlinks?: boolean;
  /** Require path to exist (default: false) */
  mustExist?: boolean;
  /** Allowed file extensions (e.g., [".txt", ".json"]) */
  allowedExtensions?: string[];
  /** Disallowed filename patterns (regex strings) */
  disallowedPatterns?: string[];
  /** Custom error message prefix */
  errorPrefix?: string;
}

/**
 * Custom error class for path validation failures
 */
export class PathValidationError extends Error {
  constructor(
    message: string,
    public readonly attemptedPath: string,
    public readonly reason: string
  ) {
    super(message);
    this.name = "PathValidationError";
    Object.setPrototypeOf(this, PathValidationError.prototype);
  }
}

/**
 * Secure path validator to prevent directory traversal attacks
 *
 * Key security features:
 * - Rejects absolute paths outside baseDir
 * - Detects path traversal attempts (../)
 * - Optionally validates symlinks
 * - Extension and pattern filtering
 * - URL-encoding attack prevention
 *
 * @example
 * ```typescript
 * const validator = new PathValidator({ baseDir: "/workspace" });
 * const safePath = validator.validate("data/file.txt"); // OK
 * validator.validate("../etc/passwd"); // Throws PathValidationError
 * ```
 */
export class PathValidator {
  private readonly baseDir: string;
  private readonly options: Required<
    Omit<PathValidationOptions, "allowedExtensions" | "disallowedPatterns" | "errorPrefix">
  > & {
    allowedExtensions?: string[];
    disallowedPatterns?: string[];
    errorPrefix?: string;
  };

  constructor(options: PathValidationOptions) {
    // Normalize and resolve base directory
    this.baseDir = path.resolve(path.normalize(options.baseDir));
    this.options = {
      baseDir: this.baseDir,
      allowSymlinks: options.allowSymlinks ?? false,
      mustExist: options.mustExist ?? false,
      allowedExtensions: options.allowedExtensions,
      disallowedPatterns: options.disallowedPatterns,
      errorPrefix: options.errorPrefix,
    };

    // Validate base directory exists
    if (!fs.existsSync(this.baseDir)) {
      throw new Error(`Base directory does not exist: ${this.baseDir}`);
    }
  }

  /**
   * Validate a single path and return the absolute resolved path
   *
   * @param inputPath - Path to validate (relative or absolute)
   * @returns Absolute path within baseDir
   * @throws PathValidationError if validation fails
   */
  validate(inputPath: string): string {
    const prefix = this.options.errorPrefix || "Access denied";

    // Decode URL-encoded paths to prevent bypass attempts
    let decodedPath = inputPath;
    try {
      decodedPath = decodeURIComponent(inputPath);
    } catch {
      // If decode fails, use original (might be already decoded)
      decodedPath = inputPath;
    }

    // Reject paths with null bytes (directory traversal technique)
    if (decodedPath.includes("\0")) {
      throw new PathValidationError(
        `${prefix}: null byte in path`,
        inputPath,
        "null_byte"
      );
    }

    // Reject absolute paths that don't start with baseDir
    if (path.isAbsolute(decodedPath)) {
      const normalized = path.normalize(decodedPath);
      if (!normalized.startsWith(this.baseDir)) {
        throw new PathValidationError(
          `${prefix}: absolute paths outside workspace not allowed`,
          inputPath,
          "absolute_path_outside_workspace"
        );
      }
    }

    // Resolve path relative to baseDir
    const resolvedPath = path.resolve(this.baseDir, decodedPath);
    const normalizedPath = path.normalize(resolvedPath);

    // Critical: Check if resolved path escapes baseDir
    const relativePath = path.relative(this.baseDir, normalizedPath);
    const escapesWorkspace =
      relativePath.startsWith("..") || path.isAbsolute(relativePath);

    if (escapesWorkspace) {
      throw new PathValidationError(
        `${prefix}: path resolves outside workspace (attempted: ${inputPath})`,
        inputPath,
        "path_traversal"
      );
    }

    // Check if path exists (if required)
    if (this.options.mustExist && !fs.existsSync(normalizedPath)) {
      throw new PathValidationError(
        `${prefix}: path does not exist`,
        inputPath,
        "not_found"
      );
    }

    // Check for symlinks (if not allowed)
    if (!this.options.allowSymlinks && fs.existsSync(normalizedPath)) {
      try {
        const stats = fs.lstatSync(normalizedPath);
        if (stats.isSymbolicLink()) {
          throw new PathValidationError(
            `${prefix}: symlinks not allowed`,
            inputPath,
            "symlink"
          );
        }
      } catch (error) {
        if (error instanceof PathValidationError) {
          throw error;
        }
        // If lstat fails, path might not exist yet (which is OK if mustExist=false)
      }
    }

    // Check file extension (if specified)
    if (this.options.allowedExtensions) {
      const ext = path.extname(normalizedPath).toLowerCase();
      const allowed = this.options.allowedExtensions.map((e) =>
        e.toLowerCase()
      );
      if (!allowed.includes(ext)) {
        throw new PathValidationError(
          `${prefix}: file extension '${ext}' not allowed`,
          inputPath,
          "invalid_extension"
        );
      }
    }

    // Check disallowed patterns (if specified)
    if (this.options.disallowedPatterns) {
      const filename = path.basename(normalizedPath);
      for (const pattern of this.options.disallowedPatterns) {
        const regex = new RegExp(pattern);
        if (regex.test(filename)) {
          throw new PathValidationError(
            `${prefix}: filename matches disallowed pattern`,
            inputPath,
            "disallowed_pattern"
          );
        }
      }
    }

    return normalizedPath;
  }

  /**
   * Validate multiple paths in batch
   *
   * @param paths - Array of paths to validate
   * @returns Array of absolute paths
   * @throws PathValidationError if any validation fails
   */
  validateBatch(paths: string[]): string[] {
    return paths.map((p) => this.validate(p));
  }

  /**
   * Check if a path would be valid without throwing
   *
   * @param inputPath - Path to check
   * @returns true if valid, false otherwise
   */
  check(inputPath: string): boolean {
    try {
      this.validate(inputPath);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Static convenience method for one-off validation
   */
  static validate(inputPath: string, options: PathValidationOptions): string {
    const validator = new PathValidator(options);
    return validator.validate(inputPath);
  }
}

File: kb/security/path_validator.py
"""
kb/security/path_validator.py

Secure path validator to prevent directory traversal attacks.

Key security features:
- Rejects absolute paths outside base_dir
- Detects path traversal attempts (../)
- Optionally validates symlinks
- Extension and pattern filtering
- URL-encoding attack prevention
"""

import re
import unicodedata
from pathlib import Path
from urllib.parse import unquote


class PathValidationError(Exception):
    """Custom exception for path validation failures."""

    def __init__(self, message: str, attempted_path: str, reason: str):
        super().__init__(message)
        self.attempted_path = attempted_path
        self.reason = reason


class PathValidator:
    """
    Secure path validator to prevent directory traversal attacks.

    Example:
        >>> validator = PathValidator(base_dir="/workspace")
        >>> safe_path = validator.validate("data/file.txt")  # OK
        >>> validator.validate("../etc/passwd")  # Raises PathValidationError
    """

    def __init__(
        self,
        base_dir: str | Path,
        allow_symlinks: bool = False,
        must_exist: bool = False,
        allowed_extensions: list[str] | None = None,
        disallowed_patterns: list[str] | None = None,
        error_prefix: str = "Access denied",
    ):
        """
        Initialize path validator.

        Args:
            base_dir: Base directory - paths must resolve within this directory
            allow_symlinks: Allow symlinks (default: False for security)
            must_exist: Require path to exist (default: False)
            allowed_extensions: Allowed file extensions (e.g., [".txt", ".json"])
            disallowed_patterns: Disallowed filename patterns (regex strings)
            error_prefix: Custom error message prefix
        """
        self.base_dir = Path(base_dir).resolve()
        self.allow_symlinks = allow_symlinks
        self.must_exist = must_exist
        self.allowed_extensions = (
            [ext.lower() for ext in allowed_extensions] if allowed_extensions else None
        )
        self.disallowed_patterns = disallowed_patterns
        self.error_prefix = error_prefix

        # Validate base directory exists
        if not self.base_dir.exists():
            raise ValueError(f"Base directory does not exist: {self.base_dir}")

        if not self.base_dir.is_dir():
            raise ValueError(f"Base directory is not a directory: {self.base_dir}")

    def validate(self, input_path: str | Path) -> Path:
        """
        Validate a single path and return the absolute resolved path.

        Args:
            input_path: Path to validate (relative or absolute)

        Returns:
            Absolute Path object within base_dir

        Raises:
            PathValidationError: If validation fails
        """
        input_str = str(input_path)

        # Reject UNC paths (Windows network paths)
        if input_str.startswith("\\\\") or input_str.startswith("//"):
            raise PathValidationError(
                f"{self.error_prefix}: UNC paths not allowed",
                input_str,
                "unc_path",
            )

        # Normalize backslashes to forward slashes
        normalized = input_str.replace("\\", "/")

        # Unicode normalization to prevent homoglyph attacks
        normalized = unicodedata.normalize("NFKC", normalized)

        # Check for suspicious Unicode characters that might be used in attacks
        # Unicode dot leader (U+2024), for example, should not be in file paths
        suspicious_chars = ["\u2024", "\u2025", "\u2026"]  # Various dot characters
        for char in suspicious_chars:
            if char in normalized:
                raise PathValidationError(
                    f"{self.error_prefix}: suspicious Unicode character in path",
                    input_str,
                    "suspicious_unicode",
                )

        # Decode URL-encoded paths multiple times to prevent double encoding bypass
        decoded_path = normalized
        for _ in range(3):  # Decode up to 3 times to catch double/triple encoding
            try:
                prev = decoded_path
                decoded_path = unquote(decoded_path)
                if decoded_path == prev:
                    break  # No more decoding needed
            except Exception:
                break

        # Reject paths with null bytes (directory traversal technique)
        if "\0" in decoded_path:
            raise PathValidationError(
                f"{self.error_prefix}: null byte in path",
                input_str,
                "null_byte",
            )

        # Convert to Path object
        path_obj = Path(decoded_path)

        # Reject absolute paths that don't start with base_dir
        if path_obj.is_absolute():
            try:
                # Check if absolute path is within workspace
                path_obj.relative_to(self.base_dir)
            except ValueError:
                raise PathValidationError(
                    f"{self.error_prefix}: absolute paths outside workspace not allowed",
                    input_str,
                    "absolute_path_outside_workspace",
                )

        # Resolve path relative to base_dir
        try:
            resolved_path = (self.base_dir / path_obj).resolve()
        except (OSError, RuntimeError) as e:
            raise PathValidationError(
                f"{self.error_prefix}: failed to resolve path: {e}",
                input_str,
                "resolution_failed",
            )

        # Critical: Check if resolved path escapes base_dir
        try:
            resolved_path.relative_to(self.base_dir)
        except ValueError:
            raise PathValidationError(
                f"{self.error_prefix}: path resolves outside workspace (attempted: {input_str})",
                input_str,
                "path_traversal",
            )

        # Check if path exists (if required)
        if self.must_exist and not resolved_path.exists():
            raise PathValidationError(
                f"{self.error_prefix}: path does not exist",
                input_str,
                "not_found",
            )

        # Check for symlinks (if not allowed)
        if not self.allow_symlinks and resolved_path.exists():
            if resolved_path.is_symlink():
                raise PathValidationError(
                    f"{self.error_prefix}: symlinks not allowed",
                    input_str,
                    "symlink",
                )

        # Check file extension (if specified)
        if self.allowed_extensions:
            ext = resolved_path.suffix.lower()
            if ext not in self.allowed_extensions:
                raise PathValidationError(
                    f"{self.error_prefix}: file extension '{ext}' not allowed",
                    input_str,
                    "invalid_extension",
                )

        # Check disallowed patterns (if specified)
        if self.disallowed_patterns:
            filename = resolved_path.name
            for pattern in self.disallowed_patterns:
                if re.match(pattern, filename):
                    raise PathValidationError(
                        f"{self.error_prefix}: filename matches disallowed pattern",
                        input_str,
                        "disallowed_pattern",
                    )

        return resolved_path

    def validate_batch(self, paths: list[str | Path]) -> list[Path]:
        """
        Validate multiple paths in batch.

        Args:
            paths: List of paths to validate

        Returns:
            List of absolute Path objects

        Raises:
            PathValidationError: If any validation fails
        """
        return [self.validate(p) for p in paths]

    def check(self, input_path: str | Path) -> bool:
        """
        Check if a path would be valid without raising an exception.

        Args:
            input_path: Path to check

        Returns:
            True if valid, False otherwise
        """
        try:
            self.validate(input_path)
            return True
        except PathValidationError:
            return False

    @staticmethod
    def validate_static(
        input_path: str | Path,
        base_dir: str | Path,
        **kwargs,
    ) -> Path:
        """
        Static convenience method for one-off validation.

        Args:
            input_path: Path to validate
            base_dir: Base directory for validation
            **kwargs: Additional options (allow_symlinks, must_exist, etc.)

        Returns:
            Absolute Path object within base_dir

        Raises:
            PathValidationError: If validation fails
        """
        validator = PathValidator(base_dir=base_dir, **kwargs)
        return validator.validate(input_path)
